nmosnode/aggregator.py

```diff
@@ -83,58 +80,288 @@ def __init__(self, logger=None, mdns_updater=None):
+                self.logger.writeDebug("Failed to find registration API")
+                break
+            self.logger.writeDebug("Aggregator set to: {}".format(self.aggregator))
+
+            if self._heartbeat_operation():
+                # Successfully registered Node
+                break
+
+    def _registered_operation(self):
+        """In Registered operation, the Node is registered so a heartbeat will be performed,
+        if the heartbeat is successful the Node will wait 5 seconds before attempting another heartbeat.
+        Else another aggregator will be selected"""
+        if not self._heartbeat_operation():
+            # Heartbeat failed
+            # Flag to update cached list of aggregators
+            self.aggregator = None
+            self._aggregator_failure = True
+
+    def _heartbeat_operation(self):
+        """Performs a heartbeat to registered aggregator
+        If heartbeat fails it will take actions to correct the error, by re-registering the Node
+        If successfull will return True, else will return False"""
+        if not self.aggregator:
+            return False
+        try:
+            self.logger.writeDebug("Sending heartbeat for Node {}"
+                                   .format(self._node_data["node"]["data"]["id"]))
+            R = self._send("POST", self.aggregator, self.aggregator_apiversion,
+                           "/health/nodes/{}".format(self._node_data["node"]["data"]["id"]))
+
+            if R.status_code == 200 and self._node_data["registered"]:
+                # Continue to registered operation
+                self._registered()
+                heartbeat_wait = 5
+                while heartbeat_wait > 0 and self._running:
+                    gevent.sleep(1)
+                    heartbeat_wait -= 1
+                return True
+
+            elif R.status_code in [200, 409]:
+                # Delete node from registry
+                if self._unregister_node(R.headers.get('Location')):
+                    return self._register_node()
+                else:
+                    # Try next registry
+                    return False
+
+        except InvalidRequest as e:
+            if e.status_code == 404:
+                # Re-register
+                self.logger.writeWarning("404 error on heartbeat. Marking Node for re-registration")
+                self._node_data["registered"] = False
+                return self._register_node()
+            else:
+                # Other error, try next registry
+                return False
+        except ServerSideError:
+            self.logger.writeWarning("Server Side Error on heartbeat. Trying another registry")
+            return False
+        except Exception:
+            # Re-register
+            self.logger.writeWarning("Unexpected error on heartbeat. Marking Node for re-registration")
+            self._node_data["registered"] = False
+            return False
+
+    def _register_node(self):
```

Comment:
It looks like this can be called from either the queue processing or main thread. Is there a danger of a race condition here? This also doesn't seem to be a noted option in the state diagram for the queue processing.

Reply:
Is it perhaps that the Node is only ever sent to the queue during updates (such as changes to its interface state)? If so perhaps the queue processing needs a separate simpler _update_node method?

Reply:
Solved by only using calling `_register_node()` on initial registration all further calls by `register()` will result in a HTTP resource request being queued. 
 Then only the main thread will call `_register_node()` in the event of a Node being unregistered from the aggregator.